pkg/etcd/etcd.go

```diff
@@ -991,25 +991,36 @@ func (e *ETCD) DeleteSnapshots(ctx context.Context, snapshots []string) error {
 
 			objectsCh := make(chan minio.ObjectInfo)
 
-			go func() {
-				for obj := range e.s3.client.ListObjects(ctx, e.config.EtcdS3BucketName, minio.ListObjectsOptions{}) {
-					if obj.Err != nil {
-						logrus.Error(obj.Err)
-						continue
-					}
-
-					// iterate through the given snapshots and only
-					// add them to the channel for remove if they're
-					// actually found from the bucket listing.
-					for _, snapshot := range snapshots {
-						if snapshot == obj.Key {
-							objectsCh <- obj
+			toCtx, cancel := context.WithTimeout(ctx, time.Second*30)
+			defer cancel()
+
+			go func(ctx context.Context) {
+				defer close(objectsCh)
+
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					default:
+						for obj := range e.s3.client.ListObjects(ctx, e.config.EtcdS3BucketName, minio.ListObjectsOptions{}) {
+							if obj.Err != nil {
+								logrus.Error(obj.Err)
+								continue
+							}
+
+							// iterate through the given snapshots and only
+							// add them to the channel for remove if they're
+							// actually found from the bucket listing.
+							for _, snapshot := range snapshots {
+								if snapshot == obj.Key {
+									objectsCh <- obj
+								}
+							}
 						}
 					}
 				}
 
-				close(objectsCh)
-			}()
+			}(toCtx)
 
 			for roErr := range e.s3.client.RemoveObjects(ctx, e.config.EtcdS3BucketName, objectsCh, minio.RemoveObjectsOptions{}) {
```

Comment:
Should make sure we've got a timeout on all the s3 ops.
```suggestion
			for roErr := range e.s3.client.RemoveObjects(toCtx, e.config.EtcdS3BucketName, objectsCh, minio.RemoveObjectsOptions{}) {
```

Reply:
This one doesn't take a context that can be used for timeout. Likely removing it anyway.